tests/Query/Filter/DueDateField.test.ts

```diff
@@ -274,29 +320,42 @@ describe('due date before relative date range (Today is 2022-05-25)', () => {
+            const filterBefore = new DueDateField().createFilterOrErrorMessage(`due before ${range}`);
+            const filterOnOrBefore = new DueDateField().createFilterOrErrorMessage(`due in or before ${range}`);
 
             // Act, Assert
-            testTaskFilterForTaskWithDueDate(filter, null, false);
-            testTaskFilterForTaskWithDueDate(filter, beforeRange, true);
-            testTaskFilterForTaskWithDueDate(filter, rangeStart, false);
-            testTaskFilterForTaskWithDueDate(filter, rangeEnd, false);
-            testTaskFilterForTaskWithDueDate(filter, afterRange, false);
+            testTaskFilterForTaskWithDueDate(filterBefore, null, false);
+            testTaskFilterForTaskWithDueDate(filterBefore, beforeRange, true);
+            testTaskFilterForTaskWithDueDate(filterBefore, rangeStart, false);
+            testTaskFilterForTaskWithDueDate(filterBefore, rangeEnd, false);
+            testTaskFilterForTaskWithDueDate(filterBefore, afterRange, false);
+
+            testTaskFilterForTaskWithDueDate(filterOnOrBefore, null, false);
+            testTaskFilterForTaskWithDueDate(filterOnOrBefore, beforeRange, true);
+            testTaskFilterForTaskWithDueDate(filterOnOrBefore, rangeStart, true);
+            testTaskFilterForTaskWithDueDate(filterOnOrBefore, rangeEnd, true);
+            testTaskFilterForTaskWithDueDate(filterOnOrBefore, afterRange, false);
         },
     );
 
     it.each([
-        ['week', '2022-05-23 (Monday 23rd May 2022)'],
-        ['month', '2022-05-01 (Sunday 1st May 2022)'],
-        ['quarter', '2022-04-01 (Friday 1st April 2022)'],
-        ['year', '2022-01-01 (Saturday 1st January 2022)'],
-    ])('should explain before relative date range (%s)', (range: string, date: string) => {
-        const filter = new DueDateField().createFilterOrErrorMessage(`due before this ${range}`);
-        expect(filter).toHaveExplanation(`due date is before ${date}`);
-    });
+        ['week', '2022-05-23 (Monday 23rd May 2022)', '2022-05-29 (Sunday 29th May 2022)'],
+        ['month', '2022-05-01 (Sunday 1st May 2022)', '2022-05-31 (Tuesday 31st May 2022)'],
+        ['quarter', '2022-04-01 (Friday 1st April 2022)', '2022-06-30 (Thursday 30th June 2022)'],
+        ['year', '2022-01-01 (Saturday 1st January 2022)', '2022-12-31 (Saturday 31st December 2022)'],
+    ])(
+        'should explain before & in or before relative date range (%s)',
+        (range: string, startDate: string, endDate: string) => {
+            const filterBefore = new DueDateField().createFilterOrErrorMessage(`due before this ${range}`);
+            expect(filterBefore).toHaveExplanation(`due date is before ${startDate}`);
+
+            const filterOnOrBefore = new DueDateField().createFilterOrErrorMessage(`due in or before this ${range}`);
+            expect(filterOnOrBefore).toHaveExplanation(`due date is on or before ${endDate}`);
```

Comment:
I had the same thought and implemented this way on purpose. Let me explain the logic behind and if you disagree at the end, I will change it.

The general idea that I got from the discussion #124 is that we use `on` when we reference a singe date (`2023-08-06`) and `in` when we reference a date range (`2023-08-02 2023-08-06`, `this month`, `2023-W32`, etc).

Now following this logic, let's consider this week `Mon 2023-07-31` - `Sun 2023-08-06`. For the instruction `due IN or before this week` (here we use `in` because we reference a range) the explanation shall be `due ON or before Sunday 2023-08-06` (here we use `on` because we reference a day). I think it is worth referencing the exact date which the filter will be using and it is easier and clearer for the user.

Or would you prefer an explanation `due in or before this week (Mon 2023-07-31 - Sun 2023-08-06)` (this is not the exact wording, just to give an idea). Or something else?

Reply:
New explanations now clarified with https://github.com/obsidian-tasks-group/obsidian-tasks/pull/2186/commits/c58aa73047361da4dbcca63b90f8162d5e270c31 =)